apps/emqx_bridge_s3tables/src/emqx_bridge_s3tables_impl.erl

```diff
@@ -0,0 +1,1117 @@
+    ConnState = ConnState0#{?installed_channels := InstalledChans},
+    {ok, ConnState};
+on_remove_channel(_ConnResId, ConnState, _ChanResId) ->
+    {ok, ConnState}.
+
+-spec on_get_channel_status(
+    connector_resource_id(),
+    action_resource_id(),
+    connector_state()
+) ->
+    ?status_connected | ?status_disconnected.
+on_get_channel_status(
+    ConnResId,
+    ChanResId,
+    _ConnState = #{?installed_channels := InstalledChans}
+) when is_map_key(ChanResId, InstalledChans) ->
+    ChanState = maps:get(ChanResId, InstalledChans),
+    channel_status(ConnResId, ChanResId, ChanState);
+on_get_channel_status(_ConnResId, _ChanResId, _ConnState) ->
+    ?status_disconnected.
+
+-spec on_query(connector_resource_id(), query(), connector_state()) ->
+    {ok, _Result} | {error, _Reason}.
+on_query(
+    _ConnResId,
+    {ChanResId, #{} = Data},
+    #{?installed_channels := InstalledChans} = _ConnState
+) when
+    is_map_key(ChanResId, InstalledChans)
+->
+    ChanState = maps:get(ChanResId, InstalledChans),
+    run_aggregated_action([Data], ChanResId, ChanState);
+on_query(_ConnResId, Query, _ConnState) ->
+    {error, {unrecoverable_error, {invalid_query, Query}}}.
+
+-spec on_batch_query(connector_resource_id(), [query()], connector_state()) ->
+    ok | {error, term()}.
+on_batch_query(
+    _ConnResId,
+    [{ChanResId, _Data} | _] = Queries,
+    #{?installed_channels := InstalledChans} = _ConnState
+) when
+    is_map_key(ChanResId, InstalledChans)
+->
+    ChanState = maps:get(ChanResId, InstalledChans),
+    Batch = [Data || {_, Data} <- Queries],
+    run_aggregated_action(Batch, ChanResId, ChanState);
+on_batch_query(_ConnResId, Queries, _ConnectorState) ->
+    {error, {unrecoverable_error, {invalid_batch, Queries}}}.
+
+%%------------------------------------------------------------------------------
+%% API
+%%------------------------------------------------------------------------------
+
+-spec load_and_memoize_schema_files() -> ok.
+load_and_memoize_schema_files() ->
+    {ok, ScJSON} = file:read_file(
+        filename:join([code:lib_dir(emqx_bridge_s3tables), "priv", "manifest-file.avsc"])
+    ),
+    Sc = avro:decode_schema(ScJSON),
+    Header = avro_ocf:make_header(Sc),
+    persistent_term:put(?MANIFEST_FILE_PT_KEY, #{schema => Sc, header => Header}),
+    %% Entry schema varies with partition spec.
+    {ok, EntryScJSONRaw} = file:read_file(
+        filename:join([code:lib_dir(emqx_bridge_s3tables), "priv", "manifest-entry.avsc"])
+    ),
+    EntryScJSON = emqx_utils_json:decode(EntryScJSONRaw),
+    persistent_term:put(?MANIFEST_ENTRY_PT_KEY, #{json_schema => EntryScJSON}),
+    ok.
+
+-spec forget_schema_files() -> ok.
+forget_schema_files() ->
+    _ = persistent_term:erase(?MANIFEST_FILE_PT_KEY),
+    _ = persistent_term:erase(?MANIFEST_ENTRY_PT_KEY),
+    ok.
+
+%%------------------------------------------------------------------------------
+%% `emqx_connector_aggreg_delivery' API
```

Comment:
Nit:
delivery implementation seems to be very loosely coupled with the connector itself, we may consider moving it into a separate module.